File: bindings/go/csa/csa.go
// Package csa provides Go bindings for libcsa via direct dynamic loading
// of the native shared library (syscall.LoadDLL + Proc.Call) rather than
// cgo -- this environment has no C compiler available to link a cgo
// build against the MSVC-built csa.dll, and LoadLibrary/GetProcAddress
// works against any DLL regardless of which compiler built it, sidestepping
// that entirely. This mirrors how the Go standard library itself calls
// arbitrary Windows DLLs (e.g. user32.dll).
//
// csa_capi.h's functions return a `csa_buffer { unsigned char* data;
// size_t size; }` struct BY VALUE. Under the Microsoft x64 calling
// convention (what the MSVC-built csa.dll uses), a 16-byte struct return
// is passed back via a HIDDEN POINTER supplied as an implicit first
// argument: the caller allocates 16 bytes of scratch space and passes its
// address as arg 0, and the callee writes {data, size} there. Every
// helper below that calls a csa_buffer-returning function passes that
// hidden pointer as the first element of the args slice.
//
// Linux/macOS support (dll_unix.go) uses purego instead of cgo, for the
// same reason: no portable C toolchain can be assumed at build time, and
// purego's dlopen/dlsym-based loading works against any .so/.dylib
// regardless of which compiler built it, exactly like this file's
// LoadLibrary/GetProcAddress approach does for Windows.
package csa

import (
	"errors"
	"path/filepath"
	"runtime"
	"unsafe"
)

// CsaError wraps a message from csa_last_error().
type CsaError struct{ msg string }

func (e *CsaError) Error() string { return e.msg }

type csaBuffer struct {
	data unsafe.Pointer
	size uintptr
}

var (
	procCompress           *lazyProc
	procDecompress         *lazyProc
	procCompressGeo2D      *lazyProc
	procCompressGeo2DLossy *lazyProc
	procDecompressGeo2D    *lazyProc
	procCompressGeo3D      *lazyProc
	procCompressGeo3DLossy *lazyProc
	procDecompressGeo3D    *lazyProc
	procCompressPose       *lazyProc
	procCompressPoseLossy  *lazyProc
	procDecompressPose     *lazyProc
	procFreeBuffer         *lazyProc
	procLastError          *lazyProc
	procCudaAvailable      *lazyProc
)

func init() {
	dll := mustLoadLibrary()
	procCompress = newLazyProc(dll, "csa_compress")
	procDecompress = newLazyProc(dll, "csa_decompress")
	procCompressGeo2D = newLazyProc(dll, "csa_compress_geo2d")
	procCompressGeo2DLossy = newLazyProc(dll, "csa_compress_geo2d_lossy")
	procDecompressGeo2D = newLazyProc(dll, "csa_decompress_geo2d")
	procCompressGeo3D = newLazyProc(dll, "csa_compress_geo3d")
	procCompressGeo3DLossy = newLazyProc(dll, "csa_compress_geo3d_lossy")
	procDecompressGeo3D = newLazyProc(dll, "csa_decompress_geo3d")
	procCompressPose = newLazyProc(dll, "csa_compress_pose")
	procCompressPoseLossy = newLazyProc(dll, "csa_compress_pose_lossy")
	procDecompressPose = newLazyProc(dll, "csa_decompress_pose")
	procFreeBuffer = newLazyProc(dll, "csa_free_buffer")
	procLastError = newLazyProc(dll, "csa_last_error")
	procCudaAvailable = newLazyProc(dll, "csa_cuda_available")
}

// libraryPath resolves the same way the Python/Rust/C# bindings do:
// CSA_LIB_PATH env var to an exact file, else ../../../build/<name>
// relative to this source file's location (captured at build time via
// runtime.Caller, the closest Go equivalent of __file__/CARGO_MANIFEST_DIR),
// where <name> is platform's native shared library name (defaultLibraryName,
// defined per-platform in dll_windows.go/dll_unix.go).
func libraryPath() (string, error) {
	if p := envOr("CSA_LIB_PATH", ""); p != "" {
		if fileExists(p) {
			return p, nil
		}
		return "", errors.New("CSA_LIB_PATH is set to " + p + " but that file does not exist")
	}
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		return "", errors.New("csa: could not determine source file location")
	}
	dir := filepath.Dir(thisFile)
	candidate := filepath.Join(dir, "..", "..", "..", "build", defaultLibraryName)
	if fileExists(candidate) {
		return candidate, nil
	}
	return "", errors.New("csa: could not locate " + defaultLibraryName + " at " + candidate +
		"; build the C++ project first (cmake --build build from the repo root), " +
		"or set CSA_LIB_PATH to the exact library path")
}

func extractAndFree(buf csaBuffer) ([]byte, error) {
	if buf.data == nil && buf.size == 0 {
		errPtr := procLastError.call()
		if errPtr != 0 {
			msg := goStringFromCStr(errPtr)
			if msg != "" {
				return nil, &CsaError{msg: msg}
			}
		}
		return []byte{}, nil
	}
	n := int(buf.size)
	result := make([]byte, n)
	src := unsafe.Slice((*byte)(buf.data), n)
	copy(result, src)
	// csa_free_buffer takes csa_buffer BY VALUE, which on x64 means the
	// caller passes a pointer to a (caller-owned) copy of the struct --
	// reuse buf itself as that copy.
	procFreeBuffer.call(uintptr(unsafe.Pointer(&buf)))
	runtime.KeepAlive(buf)
	return result, nil
}

// goStringFromCStr walks a NUL-terminated C string. `ptr` is a raw
// address handed back by syscall.Proc.Call (the r1 return value, always
// typed uintptr by Go's syscall package regardless of what it conceptually
// points to) rather than something derived from a Go unsafe.Pointer, so
// the usual "don't convert uintptr to Pointer across an expression
// boundary" rule -- which exists to stop a moving GC from invalidating a
// stale address -- doesn't actually apply here: this address refers to
// memory in libcsa's C runtime heap, which Go's GC never manages or
// moves, in the first place. `go vet` still flags this conversion
// (its heuristic can't distinguish a foreign address from a Go one);
// that's a known, reviewed false positive for this exact pattern, not
// a bug -- verified by the whole test suite's error-handling test
// actually exercising this path against the real library.
func goStringFromCStr(ptr uintptr) string {
	if ptr == 0 {
		return ""
	}
	var b []byte
	for i := 0; ; i++ {
		c := *(*byte)(unsafe.Pointer(ptr + uintptr(i))) //nolint:govet
		if c == 0 {
			break
		}
		b = append(b, c)
	}
	return string(b)
}

// Compress performs general-purpose compression: tries raw storage, the
// Pantograph Lift, and the LZ dictionary matcher, and keeps whichever
// encodes smallest. useGpu tries the CUDA path for the Pantograph Lift
// candidate if a device is available, falling back transparently otherwise.
func Compress(data []byte, useGpu bool) ([]byte, error) {
	gpu := uintptr(0)
	if useGpu {
		gpu = 1
	}
	var inputPtr uintptr
	if len(data) > 0 {
		inputPtr = uintptr(unsafe.Pointer(&data[0]))
	}
	buf := bufferReturningCall(procCompress, inputPtr, uintptr(len(data)), gpu)
	runtime.KeepAlive(data)
	return extractAndFree(buf)
}

func Decompress(data []byte) ([]byte, error) {
	var inputPtr uintptr
	if len(data) > 0 {
		inputPtr = uintptr(unsafe.Pointer(&data[0]))
	}
	buf := bufferReturningCall(procDecompress, inputPtr, uintptr(len(data)))
	runtime.KeepAlive(data)
	return extractAndFree(buf)
}

// Point2D is an integer 2D coordinate pair.
type Point2D struct{ X, Y int32 }

// Point3D is an integer 3D coordinate triple.
type Point3D struct{ X, Y, Z int32 }

func points2DToFlat(points []Point2D) []int32 {
	flat := make([]int32, len(points)*2)
	for i, p := range points {
		flat[2*i] = p.X
		flat[2*i+1] = p.Y
	}
	return flat
}

func points3DToFlat(points []Point3D) []int32 {
	flat := make([]int32, len(points)*3)
	for i, p := range points {
		flat[3*i] = p.X
		flat[3*i+1] = p.Y
		flat[3*i+2] = p.Z
	}
	return flat
}

func CompressGeo2D(points []Point2D) ([]byte, error) {
	flat := points2DToFlat(points)
	var ptr uintptr
	if len(flat) > 0 {
		ptr = uintptr(unsafe.Pointer(&flat[0]))
	}
	buf := bufferReturningCall(procCompressGeo2D, ptr, uintptr(len(points)))
	runtime.KeepAlive(flat)
	return extractAndFree(buf)
}

// CompressGeo2DLossy: quantStep <= 1 is lossless (identical to
// CompressGeo2D). resyncInterval periodically stores an exact rod to
// bound how far absolute position error can drift (0 = never).
func CompressGeo2DLossy(points []Point2D, quantStep, resyncInterval uint32) ([]byte, error) {
	flat := points2DToFlat(points)
	var ptr uintptr
	if len(flat) > 0 {
		ptr = uintptr(unsafe.Pointer(&flat[0]))
	}
	buf := bufferReturningCall(procCompressGeo2DLossy, ptr, uintptr(len(points)),
		uintptr(quantStep), uintptr(resyncInterval))
	runtime.KeepAlive(flat)
	return extractAndFree(buf)
}

// DecompressGeo2D works for both lossless and lossy blobs -- the
// quantization parameters ride in the blob itself.
func DecompressGeo2D(data []byte) ([]Point2D, error) {
	var inputPtr uintptr
	if len(data) > 0 {
		inputPtr = uintptr(unsafe.Pointer(&data[0]))
	}
	var outCount uintptr
	buf := bufferReturningCall(procDecompressGeo2D, inputPtr, uintptr(len(data)),
		uintptr(unsafe.Pointer(&outCount)))
	runtime.KeepAlive(data)
	raw, err := extractAndFree(buf)
	if err != nil {
		return nil, err
	}
	n := int(outCount)
	result := make([]Point2D, n)
	for i := 0; i < n; i++ {
		x := *(*int32)(unsafe.Pointer(&raw[i*8]))
		y := *(*int32)(unsafe.Pointer(&raw[i*8+4]))
		result[i] = Point2D{x, y}
	}
	return result, nil
}

func CompressGeo3D(points []Point3D) ([]byte, error) {
	flat := points3DToFlat(points)
	var ptr uintptr
	if len(flat) > 0 {
		ptr = uintptr(unsafe.Pointer(&flat[0]))
	}
	buf := bufferReturningCall(procCompressGeo3D, ptr, uintptr(len(points)))
	runtime.KeepAlive(flat)
	return extractAndFree(buf)
}

// CompressGeo3DLossy: same design as CompressGeo2DLossy. Tries both the
// xy+z composition and the true 3D similarity joint and keeps whichever
// encodes smaller. quantStep <= 1 is lossless (identical to CompressGeo3D).
func CompressGeo3DLossy(points []Point3D, quantStep, resyncInterval uint32) ([]byte, error) {
	flat := points3DToFlat(points)
	var ptr uintptr
	if len(flat) > 0 {
		ptr = uintptr(unsafe.Pointer(&flat[0]))
	}
	buf := bufferReturningCall(procCompressGeo3DLossy, ptr, uintptr(len(points)),
		uintptr(quantStep), uintptr(resyncInterval))
	runtime.KeepAlive(flat)
	return extractAndFree(buf)
}

func DecompressGeo3D(data []byte) ([]Point3D, error) {
	var inputPtr uintptr
	if len(data) > 0 {
		inputPtr = uintptr(unsafe.Pointer(&data[0]))
	}
	var outCount uintptr
	buf := bufferReturningCall(procDecompressGeo3D, inputPtr, uintptr(len(data)),
		uintptr(unsafe.Pointer(&outCount)))
	runtime.KeepAlive(data)
	raw, err := extractAndFree(buf)
	if err != nil {
		return nil, err
	}
	n := int(outCount)
	result := make([]Point3D, n)
	for i := 0; i < n; i++ {
		x := *(*int32)(unsafe.Pointer(&raw[i*12]))
		y := *(*int32)(unsafe.Pointer(&raw[i*12+4]))
		z := *(*int32)(unsafe.Pointer(&raw[i*12+8]))
		result[i] = Point3D{x, y, z}
	}
	return result, nil
}

// Quat is a unit-quaternion orientation (w, x, y, z).
type Quat struct{ W, X, Y, Z int32 }

// Pose is a 6-DOF sample: position, then orientation.
type Pose struct {
	Position    Point3D
	Orientation Quat
}

func posesToFlat(poses []Pose) []int32 {
	flat := make([]int32, len(poses)*7)
	for i, p := range poses {
		base := 7 * i
		flat[base] = p.Position.X
		flat[base+1] = p.Position.Y
		flat[base+2] = p.Position.Z
		flat[base+3] = p.Orientation.W
		flat[base+4] = p.Orientation.X
		flat[base+5] = p.Orientation.Y
		flat[base+6] = p.Orientation.Z
	}
	return flat
}

// CompressPose: 6-DOF pose stream -- position via the Geo3D auto-select,
// orientation via the Quaternion Joint (see DESIGN.md).
func CompressPose(poses []Pose) ([]byte, error) {
	flat := posesToFlat(poses)
	var ptr uintptr
	if len(flat) > 0 {
		ptr = uintptr(unsafe.Pointer(&flat[0]))
	}
	buf := bufferReturningCall(procCompressPose, ptr, uintptr(len(poses)))
	runtime.KeepAlive(flat)
	return extractAndFree(buf)
}

// CompressPoseLossy: posQuantStep/posResyncInterval reach the position
// half's existing lossy support; quatQuantStep/quatResyncInterval are the
// analogous knobs for the Quaternion Joint. Either quantStep <= 1 is
// lossless for that half.
func CompressPoseLossy(poses []Pose, posQuantStep, posResyncInterval, quatQuantStep, quatResyncInterval uint32) ([]byte, error) {
	flat := posesToFlat(poses)
	var ptr uintptr
	if len(flat) > 0 {
		ptr = uintptr(unsafe.Pointer(&flat[0]))
	}
	buf := bufferReturningCall(procCompressPoseLossy, ptr, uintptr(len(poses)),
		uintptr(posQuantStep), uintptr(posResyncInterval), uintptr(quatQuantStep), uintptr(quatResyncInterval))
	runtime.KeepAlive(flat)
	return extractAndFree(buf)
}

// DecompressPose works for both lossless and lossy blobs.
func DecompressPose(data []byte) ([]Pose, error) {
	var inputPtr uintptr
	if len(data) > 0 {
		inputPtr = uintptr(unsafe.Pointer(&data[0]))
	}
	var outCount uintptr
	buf := bufferReturningCall(procDecompressPose, inputPtr, uintptr(len(data)),
		uintptr(unsafe.Pointer(&outCount)))
	runtime.KeepAlive(data)
	raw, err := extractAndFree(buf)
	if err != nil {
		return nil, err
	}
	n := int(outCount)
	result := make([]Pose, n)
	for i := 0; i < n; i++ {
		base := i * 28
		x := *(*int32)(unsafe.Pointer(&raw[base]))
		y := *(*int32)(unsafe.Pointer(&raw[base+4]))
		z := *(*int32)(unsafe.Pointer(&raw[base+8]))
		qw := *(*int32)(unsafe.Pointer(&raw[base+12]))
		qx := *(*int32)(unsafe.Pointer(&raw[base+16]))
		qy := *(*int32)(unsafe.Pointer(&raw[base+20]))
		qz := *(*int32)(unsafe.Pointer(&raw[base+24]))
		result[i] = Pose{Point3D{x, y, z}, Quat{qw, qx, qy, qz}}
	}
	return result, nil
}

func CudaAvailable() bool {
	r := procCudaAvailable.call()
	return r != 0
}

File: bindings/csharp/Csa/Csa.cs
// Public C# API for libcsa, on top of the P/Invoke declarations in
// CsaNative.cs -- mirrors bindings/python/csa.py and bindings/rust/src/lib.rs
// so all three bindings agree on behavior (all three are built on the same
// C ABI in include/csa/csa_capi.h).
using System;
using System.Collections.Generic;

namespace Csa
{
    /// <summary>An error reported by libcsa, sourced from csa_last_error().</summary>
    public class CsaException : Exception
    {
        public CsaException(string message) : base(message) { }
    }

    public static class Codec
    {
        // A genuinely empty result (e.g. decompressing an empty file) looks
        // identical to a failed call at the struct level (Data == IntPtr.Zero,
        // Size == 0) -- csa_last_error() disambiguates, and it's cleared on
        // every success, same reasoning as the Python/Rust bindings.
        private static byte[] CheckAndExtract(CsaBuffer buf)
        {
            if (buf.Data == IntPtr.Zero && buf.Size == UIntPtr.Zero)
            {
                IntPtr errPtr = CsaNative.csa_last_error();
                if (errPtr != IntPtr.Zero)
                {
                    string msg = System.Runtime.InteropServices.Marshal.PtrToStringAnsi(errPtr) ?? "";
                    if (msg.Length > 0) throw new CsaException(msg);
                }
                return Array.Empty<byte>();
            }
            int size = checked((int)buf.Size);
            byte[] result = new byte[size];
            System.Runtime.InteropServices.Marshal.Copy(buf.Data, result, 0, size);
            CsaNative.csa_free_buffer(buf);
            return result;
        }

        /// <summary>
        /// General-purpose compression: tries raw storage, the Pantograph
        /// Lift, and the LZ dictionary matcher, and keeps whichever encodes
        /// smallest. useGpu tries the CUDA path for the Pantograph Lift
        /// candidate if a device is available, falling back transparently
        /// otherwise.
        /// </summary>
        public static byte[] Compress(byte[] data, bool useGpu = false)
        {
            var buf = CsaNative.csa_compress(data, (UIntPtr)data.Length, useGpu ? 1 : 0);
            return CheckAndExtract(buf);
        }

        public static byte[] Decompress(byte[] data)
        {
            var buf = CsaNative.csa_decompress(data, (UIntPtr)data.Length);
            return CheckAndExtract(buf);
        }

        private static int[] Points2DToFlat(IReadOnlyList<(int X, int Y)> points)
        {
            var flat = new int[points.Count * 2];
            for (int i = 0; i < points.Count; i++)
            {
                flat[2 * i] = points[i].X;
                flat[2 * i + 1] = points[i].Y;
            }
            return flat;
        }

        private static int[] Points3DToFlat(IReadOnlyList<(int X, int Y, int Z)> points)
        {
            var flat = new int[points.Count * 3];
            for (int i = 0; i < points.Count; i++)
            {
                flat[3 * i] = points[i].X;
                flat[3 * i + 1] = points[i].Y;
                flat[3 * i + 2] = points[i].Z;
            }
            return flat;
        }

        public static byte[] CompressGeo2D(IReadOnlyList<(int X, int Y)> points)
        {
            var flat = Points2DToFlat(points);
            var buf = CsaNative.csa_compress_geo2d(flat, (UIntPtr)points.Count);
            return CheckAndExtract(buf);
        }

        /// <summary>
        /// quantStep &lt;= 1 is lossless (identical to CompressGeo2D).
        /// resyncInterval periodically stores an exact rod to bound how far
        /// absolute position error can drift (0 = never); see DESIGN.md.
        /// </summary>
        public static byte[] CompressGeo2DLossy(IReadOnlyList<(int X, int Y)> points, uint quantStep, uint resyncInterval = 0)
        {
            var flat = Points2DToFlat(points);
            var buf = CsaNative.csa_compress_geo2d_lossy(flat, (UIntPtr)points.Count, quantStep, resyncInterval);
            return CheckAndExtract(buf);
        }

        /// <summary>
        /// Works for both lossless and lossy blobs -- the quantization
        /// parameters ride in the blob itself, there is no separate lossy
        /// decode entry point.
        /// </summary>
        public static (int X, int Y)[] DecompressGeo2D(byte[] data)
        {
            var buf = CsaNative.csa_decompress_geo2d(data, (UIntPtr)data.Length, out UIntPtr outCount);
            byte[] raw = CheckAndExtract(buf);
            int n = checked((int)outCount);
            var result = new (int X, int Y)[n];
            for (int i = 0; i < n; i++)
            {
                int x = BitConverter.ToInt32(raw, i * 8);
                int y = BitConverter.ToInt32(raw, i * 8 + 4);
                result[i] = (x, y);
            }
            return result;
        }

        public static byte[] CompressGeo3D(IReadOnlyList<(int X, int Y, int Z)> points)
        {
            var flat = Points3DToFlat(points);
            var buf = CsaNative.csa_compress_geo3d(flat, (UIntPtr)points.Count);
            return CheckAndExtract(buf);
        }

        /// <summary>
        /// Same design as CompressGeo2DLossy. Tries both the xy+z
        /// composition and the true 3D similarity joint and keeps whichever
        /// encodes smaller. quantStep &lt;= 1 is lossless (identical to CompressGeo3D).
        /// </summary>
        public static byte[] CompressGeo3DLossy(IReadOnlyList<(int X, int Y, int Z)> points, uint quantStep, uint resyncInterval = 0)
        {
            var flat = Points3DToFlat(points);
            var buf = CsaNative.csa_compress_geo3d_lossy(flat, (UIntPtr)points.Count, quantStep, resyncInterval);
            return CheckAndExtract(buf);
        }

        public static (int X, int Y, int Z)[] DecompressGeo3D(byte[] data)
        {
            var buf = CsaNative.csa_decompress_geo3d(data, (UIntPtr)data.Length, out UIntPtr outCount);
            byte[] raw = CheckAndExtract(buf);
            int n = checked((int)outCount);
            var result = new (int X, int Y, int Z)[n];
            for (int i = 0; i < n; i++)
            {
                int x = BitConverter.ToInt32(raw, i * 12);
                int y = BitConverter.ToInt32(raw, i * 12 + 4);
                int z = BitConverter.ToInt32(raw, i * 12 + 8);
                result[i] = (x, y, z);
            }
            return result;
        }

        /// <summary>A 6-DOF pose sample: position, then a unit-quaternion orientation.</summary>
        public struct Pose
        {
            public (int X, int Y, int Z) Position;
            public (int W, int X, int Y, int Z) Orientation;
            public Pose((int, int, int) position, (int, int, int, int) orientation)
            {
                Position = position;
                Orientation = orientation;
            }
        }

        private static int[] PosesToFlat(IReadOnlyList<Pose> poses)
        {
            var flat = new int[poses.Count * 7];
            for (int i = 0; i < poses.Count; i++)
            {
                int b = 7 * i;
                flat[b] = poses[i].Position.X;
                flat[b + 1] = poses[i].Position.Y;
                flat[b + 2] = poses[i].Position.Z;
                flat[b + 3] = poses[i].Orientation.W;
                flat[b + 4] = poses[i].Orientation.X;
                flat[b + 5] = poses[i].Orientation.Y;
                flat[b + 6] = poses[i].Orientation.Z;
            }
            return flat;
        }

        /// <summary>
        /// 6-DOF pose stream: position via the Geo3D auto-select,
        /// orientation via the Quaternion Joint (see DESIGN.md).
        /// </summary>
        public static byte[] CompressPose(IReadOnlyList<Pose> poses)
        {
            var flat = PosesToFlat(poses);
            var buf = CsaNative.csa_compress_pose(flat, (UIntPtr)poses.Count);
            return CheckAndExtract(buf);
        }

        /// <summary>
        /// posQuantStep/posResyncInterval reach the position half's existing
        /// lossy support; quatQuantStep/quatResyncInterval are the analogous
        /// knobs for the Quaternion Joint. Either quantStep &lt;= 1 is
        /// lossless for that half.
        /// </summary>
        public static byte[] CompressPoseLossy(IReadOnlyList<Pose> poses,
            uint posQuantStep, uint posResyncInterval, uint quatQuantStep, uint quatResyncInterval)
        {
            var flat = PosesToFlat(poses);
            var buf = CsaNative.csa_compress_pose_lossy(flat, (UIntPtr)poses.Count,
                posQuantStep, posResyncInterval, quatQuantStep, quatResyncInterval);
            return CheckAndExtract(buf);
        }

        /// <summary>Works for both lossless and lossy blobs.</summary>
        public static Pose[] DecompressPose(byte[] data)
        {
            var buf = CsaNative.csa_decompress_pose(data, (UIntPtr)data.Length, out UIntPtr outCount);
            byte[] raw = CheckAndExtract(buf);
            int n = checked((int)outCount);
            var result = new Pose[n];
            for (int i = 0; i < n; i++)
            {
                int b = i * 28;
                int x = BitConverter.ToInt32(raw, b);
                int y = BitConverter.ToInt32(raw, b + 4);
                int z = BitConverter.ToInt32(raw, b + 8);
                int qw = BitConverter.ToInt32(raw, b + 12);
                int qx = BitConverter.ToInt32(raw, b + 16);
                int qy = BitConverter.ToInt32(raw, b + 20);
                int qz = BitConverter.ToInt32(raw, b + 24);
                result[i] = new Pose((x, y, z), (qw, qx, qy, qz));
            }
            return result;
        }

        public static bool CudaAvailable() => CsaNative.csa_cuda_available() != 0;
    }
}
